File: freighter/py/freighter/endpoint.py
from __future__ import annotations
from urllib.parse import urljoin
from functools import reduce


class Endpoint:
    protocol: str
    host: str
    port: int
    path_prefix: str

    def __init__(
            self, host: str, port: int, path_prefix: str = "", protocol: str = ""
    ) -> None:
        self.protocol = protocol
        self.host = host
        self.port = port
        self.path_prefix = format_path(path_prefix)

    def path(self, path: str) -> str:
        return reduce(urljoin, [self.uri(), self.path_prefix, format_path(path)])

    def child(self, path: str, protocol: str = "") -> Endpoint:
        if protocol == "":
            protocol = self.protocol
        return Endpoint(self.host, self.port, self._child_prefix(path), protocol)

    def uri(self) -> str:
        return f"{self.protocol}://{self.host}:{self.port}"

    def _child_prefix(self, path: str):
        return reduce(urljoin, [self.path_prefix, format_path(path)])

    def __str__(self) -> str:
        return self.uri()


def format_path(path: str):
    path = path if path.endswith("/") else f"{path}/"
    path = path[1:] if path.startswith("/") else path
    return path

File: freighter/ts/src/lib/endpoint.ts
type EndpointProps = {
  host: string;
  port: number;
  protocol?: string;
  pathPrefix?: string;
};

export default class Endpoint {
  protocol: string;
  host: string;
  port: number;
  pathPrefix: string;

  constructor({ host, port, protocol = '', pathPrefix = '' }: EndpointProps) {
    this.protocol = protocol;
    this.host = host;
    this.port = port;
    this.pathPrefix = pathPrefix;
  }

  child({
    path,
    protocol = '',
  }: {
    path?: string;
    protocol?: string;
  }): Endpoint {
    return new Endpoint({
      host: this.host,
      port: this.port,
      protocol: protocol || this.protocol,
      pathPrefix: path ? joinPath(this.pathPrefix, path) : this.pathPrefix,
    });
  }

  path(path: string): string {
    return joinPath(this.uri(), path);
  }

  uri(): string {
    return joinPath(
      `${this.protocol}://${this.host}:${this.port}`,
      this.pathPrefix
    );
  }
}

// joinPath joins the two paths, ensuring there is a single slash between them.
const joinPath = (a: string, b: string): string => {
  // Remove any leading slashes from b.
  b = b.replace(/^\/+/, '');
  // Remove any trailing slashes from a.
  a = a.replace(/\/+$/, '');
  // Join the two paths with a single slash.
  return `${a}/${b}`;
};
